### PR TITLE
No failure reason displayed when build failed using invalid contextDir

### PR DESCRIPTION
**Pull request by coreydaley**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1419810

### CODE REVIEW

**Comment by coreydaley** [comment]
[test]

**Comment by coreydaley** [comment]
@openshift/devex ptal

**Comment by bparees** [comment]
@coreydaley what about s2i flows where the specified context dir doesn't exist?  it seems like we could/should be making this check in a more common/shared location, where we are fetching the source.

**Comment by coreydaley** [comment]
@bparees Refactored to do the check in fetchSource, covers both docker and s2i now.

File: pkg/build/builder/docker.go [hunk 1]
```diff
@@ -68,9 +68,17 @@ func (d *DockerBuilder) Build() error {
 	}
 	sourceInfo, err := fetchSource(d.dockerClient, buildDir, d.build, initialURLCheckTimeout, os.Stdin, d.gitClient)
 	if err != nil {
-		d.build.Status.Phase = api.BuildPhaseFailed
-		d.build.Status.Reason = api.StatusReasonFetchSourceFailed
-		d.build.Status.Message = api.StatusMessageFetchSourceFailed
+		switch {
+		case strings.Contains(err.Error(), "no such file or directory"):
```
Comment (bparees) [inline pkg/build/builder/docker.go]: can't you just check if it's an os.IsNotExist error like you did in fetchSource?

Comment (bparees) [inline pkg/build/builder/docker.go]: and better still would probably be to wrap the error inside fetchSource and check for that explicit type.  that way if fetchsource throws a file not found error for another reason, you won't report the wrong reason.

**Comment by coreydaley** [comment]
@bparees updated with your suggestion

File: pkg/build/builder/source.go [hunk 2]
```diff
@@ -92,6 +97,12 @@ func fetchSource(dockerClient DockerClient, dir string, build *api.Build, urlTim
 		}
 	}
 
+	if len(build.Spec.Source.ContextDir) > 0 {
+		if _, err := os.Stat(filepath.Join(dir, build.Spec.Source.ContextDir)); os.IsNotExist(err) {
+			return sourceInfo, contextDirNotFoundError(filepath.Join(dir, build.Spec.Source.ContextDir))
```
Comment (bparees) [inline pkg/build/builder/source.go]: just report the contextdir path, "dir" is going to be an internal path with no real meaning to someone who sees it.

Comment (coreydaley) [inline pkg/build/builder/source.go]: Fixed.

**Comment by bparees** [comment]
lgtm, please create another pull for release-1.5 and also a follow up to add automated tests.
[merge]

**Comment by bparees** [comment]
[testextended][extended:core(builds)]

**Comment by bparees** [comment]
[merge]

**Pull request merged**